### PR TITLE
alibabacloud: use NextToken and MaxResults to list instance types

### PR DESCRIPTION
**Pull request by haozhangami**
Currently, there is no paging query used to retrieve instance types. When there are many instance types, some instances can not obtain ENI and IPPerENI limits. The cilium operator log is as follows.
```
level=warning msg="Unable to maintain ip pool of node" error="Unable to determine limits" instanceID=<ID> name=<NAME> subsys=ipam
```
This patch fixes this problem by using NextToken and MaxResults for paging query.

```release-note
Fix bug in AlibabaCloud where instance type limits could not be determined
```

### CODE REVIEW

File: pkg/alibabacloud/api/api.go [hunk 1]
```diff
@@ -182,13 +182,26 @@ func (c *Client) GetVPCs(ctx context.Context) (ipamTypes.VirtualNetworkMap, erro
 
 // GetInstanceTypes returns all the known ECS instance types in the configured region
 func (c *Client) GetInstanceTypes(ctx context.Context) ([]ecs.InstanceType, error) {
+	var result []ecs.InstanceType
 	req := ecs.CreateDescribeInstanceTypesRequest()
-	resp, err := c.ecsClient.DescribeInstanceTypes(req)
-	if err != nil {
-		return nil, err
+	req.MaxResults = requests.NewInteger(100)
```
Comment (christarazi) [inline pkg/alibabacloud/api/api.go:191]: Why was 100 chosen? It would be good to have a brief explanation.

**Review by christarazi — COMMENTED** [review]
Thanks for the PR!

I've left one comment and I have an overarching question: how does this change resolve the error you experienced? I'm not sure I follow why this fixes it.

File: pkg/alibabacloud/api/api.go (hunk 1 above)
Comment (haozhangami) [reply pkg/alibabacloud/api/api.go:191]: Thank you for your review.

The maximum number of entries per page is 100, I add comments.

When there are many instance types, some instance type limits can not be queried through the current method, and this problem can be fixed by paging query.

Alibaba Cloud API Reference
[1] https://www.alibabacloud.com/help/en/elastic-compute-service/latest/describeinstancetypes

Comment (christarazi) [reply pkg/alibabacloud/api/api.go:191]: Thanks for the explanation. Could you add the above and the link to the reference as a code comment as well?

Comment (haozhangami) [reply pkg/alibabacloud/api/api.go:191]: Done

**Review by christarazi — APPROVED** [review]

**Comment by haozhangami** [comment]
Hi @christarazi @sayboras , please help to see if this PR can be merged.

**Comment by christarazi** [comment]
@haozhangami Could you rebase on PR on latest `main` so we can re-run the CI?

**Comment by haozhangami** [comment]
> @haozhangami Could you rebase on PR on latest `main` so we can re-run the CI?

Done

**Pull request merged**